File: built-in plugins dsp/mb-synth-poly.cpp
/**
 * MB Poly Synth
 * Category : instrument
 * Type     : analog
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Polyphonic synth
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SYNTH_POLY_H
#define MB_SYNTH_POLY_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSynthPoly : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-synth-poly";
    static constexpr const char* PLUGIN_NAME    = "MB Poly Synth";
    static constexpr const char* PLUGIN_TYPE    = "analog";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float spread = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbSynthPoly() = default;
    ~MbSynthPoly() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.spread = std::clamp(params.spread, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Poly Synth
        return input;
    }
};

#endif // MB_SYNTH_POLY_H

File: built-in plugins dsp/mb-synth-poly.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSynthPolyPlugin: PluginDefinition = {
    id: 'mb-synth-poly', slug: 'mb-synth-poly', name: 'MB Poly Synth', category: 'instrument', type: 'analog', version: '1.0.0',
    description: 'Polyphonic synth', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'sawtooth', detune: -3, gain: 0.35 }, { type: 'sawtooth', detune: 3, gain: 0.35 }, { type: 'triangle', detune: 0, gain: 0.3 }],
    envelope: { attack: 0.02, decay: 0.4, sustain: 0.6, release: 0.5 },
    parameters: [
      { id: 'spread', name: 'Spread', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { spread: 0.5, volume: 0.8 },
  };

export default MbSynthPolyPlugin;
